File: UVITRepo/UVITRepo.swift
//
//  UVITRepo.swift
//  UVITRepo
//
//  Munki 7 repo plugin — pushes packages into UVIT via the /api/repo HTTP API.
//
//  MIT License — Copyright (c) 2026 EigerCode GmbH
//  See LICENSE for the full license text.

import Foundation

// MARK: - Error types

class RepoError: Error, CustomStringConvertible {
    private let message: String

    public init(_ message: String) {
        self.message = message
    }

    public var description: String { message }
}

extension RepoError: LocalizedError {
    var errorDescription: String? { message }
}

class HTTPRepoError: Error, CustomStringConvertible {
    public let statusCode: Int
    public let body: String

    public init(_ statusCode: Int, body: String = "") {
        self.statusCode = statusCode
        self.body = body
    }

    public var description: String {
        switch statusCode {
        case 401:
            return "HTTP 401 Unauthorized — check that UVIT_TOKEN is set and not expired"
        case 403:
            return "HTTP 403 Forbidden — token lacks permission for this tenant/repo"
        default:
            let suffix = body.isEmpty ? "" : ": \(body.prefix(200))"
            return "HTTP error \(statusCode)\(suffix)"
        }
    }
}

extension HTTPRepoError: LocalizedError {
    var errorDescription: String? { description }
}

// MARK: - HTTP helpers

/// Throws an HTTPRepoError if the HTTP status code is not 2xx.
private func throwIfStatusNotOK(_ response: URLResponse, body: Data = Data()) throws {
    guard let http = response as? HTTPURLResponse else {
        throw RepoError("Response was not an HTTPURLResponse")
    }
    guard (200...299).contains(http.statusCode) else {
        let bodyString = String(data: body.prefix(512), encoding: .utf8) ?? ""
        throw HTTPRepoError(http.statusCode, body: bodyString)
    }
}

// MARK: - UVITRepo

/// Munki 7 repo plugin that targets the UVIT /api/repo HTTP ingest API.
///
/// ## Configuration
///
/// Read from the macOS preferences domain `ch.eigercode.uvit.munkiimport`, with
/// environment variable fallbacks:
///
/// | Pref key      | Env var       | Description                                           |
/// |---------------|---------------|-------------------------------------------------------|
/// | `uvitToken`   | `UVIT_TOKEN`  | Bearer JWT minted from the UVIT console (required).   |
/// |               |               | The JWT already encodes tenant + repo binding.        |
/// | `uvitTarget`  | `UVIT_TARGET` | Optional target hint (`global` or `tenant:<id>`).     |
/// |               |               | Supplementary — the JWT carries the binding.          |
/// | `uvitRepoID`  | `UVIT_REPO_ID`| Optional repo-ID hint. Same note as uvitTarget.       |
///
/// Write system-wide prefs (requires sudo):
///
///   sudo defaults write /Library/Preferences/ch.eigercode.uvit.munkiimport \
///     uvitToken "eyJ..."
///
/// ## Endpoint mapping
///
/// | Munki call                              | UVIT API call                              |
/// |-----------------------------------------|--------------------------------------------|
/// | `list("pkgsinfo")`                      | `GET <baseURL>/pkgsinfo`                   |
/// | `list(<other>)`                         | returns [] (server-side follow-up needed)  |
/// | `get("pkgsinfo/<name>/<ver>.plist")`    | `GET <baseURL>/pkgsinfo/<name>/<ver>.plist`|
/// | `get("pkgs/<path>")`                    | `GET <baseURL>/pkgs/<path>` (307 redirect) |
/// | `put("pkgs/<path>", fromFile:)`         | `PUT <baseURL>/pkgs/<path>`                |
/// | `put("pkgsinfo/<path>", content:)`      | `PUT <baseURL>/pkgsinfo/<path>`            |
/// | `delete("pkgsinfo/<path>")`             | `DELETE <baseURL>/pkgsinfo/<path>`         |
/// | `delete("pkgs/<path>")`                 | `DELETE <baseURL>/pkgs/<path>`             |
/// | `pathFor(_)`                            | returns nil (non-filesystem repo)          |
///
/// Note: UVIT catalogs are dynamic; makecatalogs is a no-op on the server side.
// ponytail: server-side list endpoints for catalogs/manifests/icons are needed
// for full makecatalogs + iconimporter parity; tracked as a server-side follow-up.
// When the server adds them, extend list() to handle those kinds here.
class UVITRepo: Repo {
    static let prefsDomain = "ch.eigercode.uvit.munkiimport"

    var baseURL: URL
    var token: String = ""
    var target: String = ""
    var repoID: String = ""

    required init(_ url: String) throws {
        guard let parsedURL = URL(string: url) else {
            throw RepoError("Could not create valid URL from \(url)")
        }
        self.baseURL = parsedURL
        loadCredentials()
    }

    // MARK: - Credential loading

    /// Loads credentials from the macOS preferences domain
    /// `ch.eigercode.uvit.munkiimport`, falling back to environment variables.
    private func loadCredentials() {
        let prefs = UserDefaults(suiteName: Self.prefsDomain)
        let env = ProcessInfo.processInfo.environment

        token = prefs?.string(forKey: "uvitToken") ?? env["UVIT_TOKEN"] ?? ""
        target = prefs?.string(forKey: "uvitTarget") ?? env["UVIT_TARGET"] ?? ""
        repoID = prefs?.string(forKey: "uvitRepoID") ?? env["UVIT_REPO_ID"] ?? ""

        if token.isEmpty {
            fputs("[UVITRepo] WARNING: No token found in preferences or environment.\n", stderr)
            fputs("[UVITRepo] Set uvitToken in /Library/Preferences/\(Self.prefsDomain).plist\n", stderr)
            fputs("[UVITRepo] or export UVIT_TOKEN=<token> before running munkiimport.\n", stderr)
        }
    }

    // MARK: - Request builder

    /// Returns a URLRequest with auth headers applied.
    ///
    /// - Parameter includeRepoID: Pass `true` for mutating requests
    ///   (PUT/DELETE on pkgs and pkgsinfo). Adds `X-UVIT-Repo-ID` if configured.
    private func baseRequest(_ url: URL, includeRepoID: Bool = false) -> URLRequest {
        var request = URLRequest(url: url)
        if !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if !target.isEmpty {
            request.setValue(target, forHTTPHeaderField: "X-UVIT-Target")
        }
        if includeRepoID && !repoID.isEmpty {
            request.setValue(repoID, forHTTPHeaderField: "X-UVIT-Repo-ID")
        }
        return request
    }

    /// Appends a path component to the base URL.
    private func url(appending path: String) -> URL {
        if #available(macOS 13.0, *) {
            return baseURL.appending(path: path)
        } else {
            return baseURL.appendingPathComponent(path)
        }
    }

    // MARK: - Repo protocol

    /// Lists item paths for the given repo kind.
    ///
    /// Only `pkgsinfo` has a server-side list endpoint today. Other kinds
    /// (catalogs, manifests, icons) return an empty list so that makecatalogs
    /// and iconimporter degrade gracefully rather than erroring out.
    func list(_ kind: String) async throws -> [String] {
        let resourceType = (kind as NSString).pathComponents.first ?? kind

        // Only pkgsinfo is implemented server-side; other kinds need follow-up.
        guard resourceType == "pkgsinfo" else {
            return []
        }

        let listURL = url(appending: "pkgsinfo")
        var request = baseRequest(listURL)
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await URLSession.shared.data(for: request)
        try throwIfStatusNotOK(response, body: data)

        // Server response: {"items": [{"path": "Name/1.0.plist"}, ...]}
        struct ListItem: Decodable { let path: String }
        struct ListResponse: Decodable { let items: [ListItem] }

        let decoded = try JSONDecoder().decode(ListResponse.self, from: data)
        return decoded.items.map(\.path)
    }

    /// Returns the raw bytes of a repo item.
    ///
    /// For `pkgsinfo/*` this is the plist blob. For `pkgs/*` the server issues
    /// a 307 redirect to a presigned S3 URL; URLSession follows it automatically.
    func get(_ identifier: String) async throws -> Data {
        let itemURL = url(appending: identifier)
        let request = baseRequest(itemURL)
        let (data, response) = try await URLSession.shared.data(for: request)
        try throwIfStatusNotOK(response, body: data)
        return data
    }

    /// Downloads a repo item and writes it to `local_file_path`.
    func get(_ identifier: String, toFile local_file_path: String) async throws {
        let data = try await get(identifier)
        FileManager.default.createFile(atPath: local_file_path, contents: data)
    }

    /// Uploads in-memory `content` to the repo at `identifier`.
    ///
    /// Use for pkgsinfo blobs. For large installer binaries prefer `put(_:fromFile:)`.
    func put(_ identifier: String, content: Data) async throws {
        let itemURL = url(appending: identifier)
        var request = baseRequest(itemURL, includeRepoID: true)
        request.httpMethod = "PUT"
        let resourceType = (identifier as NSString).pathComponents.first ?? ""
        if resourceType == "pkgsinfo" {
            request.setValue("application/xml", forHTTPHeaderField: "Content-Type")
        } else {
            request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        }
        let (responseData, response) = try await URLSession.shared.upload(for: request, from: content)
        try throwIfStatusNotOK(response, body: responseData)
    }

    /// Streams `local_file_path` to the repo at `identifier`.
    ///
    /// Use for installer binaries (pkgs/*). URLSession streams the file without
    /// loading it entirely into memory.
    func put(_ identifier: String, fromFile local_file_path: String) async throws {
        let itemURL = url(appending: identifier)
        var request = baseRequest(itemURL, includeRepoID: true)
        request.httpMethod = "PUT"
        let resourceType = (identifier as NSString).pathComponents.first ?? ""
        if resourceType == "pkgsinfo" {
            request.setValue("application/xml", forHTTPHeaderField: "Content-Type")
        } else {
            request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        }
        let localFile = URL(fileURLWithPath: local_file_path)
        let (responseData, response) = try await URLSession.shared.upload(for: request, fromFile: localFile)
        try throwIfStatusNotOK(response, body: responseData)
    }

    /// Deletes a repo item.
    func delete(_ identifier: String) async throws {
        let itemURL = url(appending: identifier)
        var request = baseRequest(itemURL, includeRepoID: true)
        request.httpMethod = "DELETE"
        let (responseData, response) = try await URLSession.shared.data(for: request)
        try throwIfStatusNotOK(response, body: responseData)
    }

    /// Non-filesystem repo — no local path exists for any identifier.
    func pathFor(_: String) -> String? { nil }
}

// MARK: - dylib entry point

/// C-callable factory function loaded by Munki's plugin system.
///
/// Munki resolves `createPlugin` from the dylib at runtime; this function
/// instantiates a `UVITRepoBuilder` and hands it to the Objective-C runtime.
@_cdecl("createPlugin")
public func createPlugin() -> UnsafeMutableRawPointer {
    return Unmanaged.passRetained(UVITRepoBuilder()).toOpaque()
}

final class UVITRepoBuilder: RepoPluginBuilder {
    override func connect(_ url: String) -> Repo? {
        return try? UVITRepo(url)
    }
}

File: UVITRepo.py
# encoding: utf-8
#
# UVITRepo.py
#
# Munki *Python* repo plugin — pushes packages into UVIT via the /api/repo
# HTTP API. This is the twin of UVITRepo.swift (the Munki 7 Swift plugin):
# same endpoints, same headers, same configuration.
#
# Why two plugins: Munki 7's Swift tools (munkiimport etc.) load Swift dylib
# plugins from /usr/local/munki/repoplugins/, while AutoPkg's MunkiImporter
# processor loads Python plugins via munkilib.munkirepo (still shipped by
# Munki 7.x as the Munki 6.7 compatibility component). AutoPkg therefore
# needs this file, installed at /usr/local/munki/munkilib/munkirepo/UVITRepo.py.
#
# HTTP goes through /usr/bin/curl (like Munki's MWA2APIRepo): AutoPkg's and
# Munki's bundled Pythons ship without a usable CA bundle, so urllib fails
# TLS verification; curl uses the system trust store.
#
# Configuration (prefs domain ch.eigercode.uvit.munkiimport, env fallback):
#   uvitToken  / UVIT_TOKEN    — required; opaque API token (uvit_pat_...)
#   uvitTarget / UVIT_TARGET   — required; "global" or "tenant:<id>"
#   uvitRepoID / UVIT_REPO_ID  — required for writes; numeric Software Repo ID
#
# MIT License — Copyright (c) 2026 EigerCode GmbH
# See LICENSE for the full license text.

from __future__ import absolute_import, print_function

import json
import os
import subprocess
import sys
import tempfile
import urllib.parse

from munkilib.munkirepo import Repo, RepoError

PREFS_DOMAIN = "ch.eigercode.uvit.munkiimport"

CURL = "/usr/bin/curl"
CONNECT_TIMEOUT = "30"


def _pref(key):
    """Reads a value from the ch.eigercode.uvit.munkiimport prefs domain.

    Returns None if PyObjC is unavailable (e.g. a stock CI Python) so the
    caller falls through to environment variables."""
    try:
        from Foundation import CFPreferencesCopyAppValue
    except ImportError:
        return None
    value = CFPreferencesCopyAppValue(key, PREFS_DOMAIN)
    if value is None:
        return None
    return str(value)


def _parse_header_dump(path):
    """Returns (status_code, headers_dict) from a curl --dump-header file.

    The dump may contain several blocks (e.g. '100 Continue' before the
    final response); the last status line and its headers win."""
    status = 0
    headers = {}
    with open(path, encoding="utf-8", errors="replace") as fileobj:
        for line in fileobj:
            line = line.strip()
            if line.upper().startswith("HTTP/"):
                try:
                    status = int(line.split()[1])
                except (IndexError, ValueError):
                    status = 0
                headers = {}
            elif ":" in line:
                key, _, value = line.partition(":")
                headers[key.strip().lower()] = value.strip()
    return status, headers


class UVITRepo(Repo):
    """Munki repo plugin backed by the UVIT /api/repo HTTP ingest API."""

    def __init__(self, baseurl):
        self.baseurl = baseurl.rstrip("/")
        self.token = _pref("uvitToken") or os.environ.get("UVIT_TOKEN", "")
        self.target = _pref("uvitTarget") or os.environ.get("UVIT_TARGET", "")
        self.repo_id = _pref("uvitRepoID") or os.environ.get("UVIT_REPO_ID", "")

        if not self.token:
            print(
                "[UVITRepo] WARNING: UVIT_TOKEN not found. Set uvitToken in\n"
                "[UVITRepo]   /Library/Preferences/%s.plist\n"
                "[UVITRepo] or export UVIT_TOKEN=uvit_pat_..." % PREFS_DOMAIN,
                file=sys.stderr,
            )
        if not self.target:
            # The server rejects requests without a target with 400.
            raise RepoError(
                "UVIT_TARGET is not set. Set uvitTarget in "
                "/Library/Preferences/%s.plist or export UVIT_TARGET=global "
                "(or tenant:<id>)." % PREFS_DOMAIN
            )

    # -- HTTP via curl -----------------------------------------------------

    def _run_curl(self, url, method="GET", headers=None, upload_file=None,
                  output_file=None):
        """Performs one HTTP request. Returns (status, headers, body).

        Redirects are NOT followed here: curl forwards custom -H headers to
        redirect targets, which would leak the bearer token (and break S3
        presigned auth). Callers follow redirects explicitly.

        body is None when output_file is given (curl streams to it).
        Request headers go through a temp file (--header @file) so the
        token never appears in the process list."""
        with tempfile.TemporaryDirectory() as tmpdir:
            header_dump = os.path.join(tmpdir, "response_headers")
            header_file = os.path.join(tmpdir, "request_headers")
            body_path = output_file or os.path.join(tmpdir, "body")

            with open(header_file, "w", encoding="utf-8") as fileobj:
                for key, value in (headers or {}).items():
                    fileobj.write("%s: %s\n" % (key, value))

            cmd = [
                CURL, "--silent", "--show-error",
                "--connect-timeout", CONNECT_TIMEOUT,
                "--dump-header", header_dump,
                "--header", "@%s" % header_file,
                "--output", body_path,
            ]
            if upload_file:
                # --upload-file streams the file and implies PUT.
                cmd.extend(["--upload-file", upload_file])
            elif method != "GET":
                cmd.extend(["--request", method])
            cmd.append(url)

            try:
                proc = subprocess.run(
                    cmd, capture_output=True, text=True, check=False)
            except OSError as err:
                raise RepoError("Could not run curl: %s" % err) from err
            if proc.returncode != 0:
                raise RepoError(
                    "Connection error: %s" % proc.stderr.strip())

            status, response_headers = _parse_header_dump(header_dump)
            body = None
            if not output_file:
                with open(body_path, "rb") as fileobj:
                    body = fileobj.read()
            return status, response_headers, body

    def _auth_headers(self, include_repo_id=False, content_type=None,
                      accept=None):
        headers = {"X-UVIT-Target": self.target}
        if self.token:
            headers["Authorization"] = "Bearer %s" % self.token
        if include_repo_id and self.repo_id:
            headers["X-UVIT-Repo-ID"] = self.repo_id
        if content_type:
            headers["Content-Type"] = content_type
        if accept:
            headers["Accept"] = accept
        return headers

    def _request(self, identifier, method="GET", upload_file=None,
                 output_file=None, include_repo_id=False, accept=None):
        """Issues a request against <baseurl>/<identifier>.

        GETs answered with a redirect (pkgs/* → presigned S3) are followed
        with a second, credential-free request."""
        url = self.baseurl + "/" + urllib.parse.quote(identifier)
        content_type = None
        if upload_file:
            resource_type = identifier.split("/", 1)[0]
            content_type = ("application/xml" if resource_type == "pkgsinfo"
                            else "application/octet-stream")
        headers = self._auth_headers(include_repo_id, content_type, accept)

        status, response_headers, body = self._run_curl(
            url, method, headers, upload_file, output_file)

        if method == "GET" and 300 <= status <= 399:
            location = response_headers.get("location")
            if not location:
                raise RepoError("HTTP %s redirect without Location" % status)
            status, response_headers, body = self._run_curl(
                location, "GET", {}, None, output_file)

        if status == 401:
            raise RepoError(
                "HTTP 401 Unauthorized — check that UVIT_TOKEN "
                "(uvit_pat_…) is set and valid")
        if status == 403:
            raise RepoError(
                "HTTP 403 Forbidden — token owner is not authorized "
                "for the target tenant/repo")
        if not 200 <= status <= 299:
            detail = ""
            if body:
                detail = body[:200].decode("utf-8", errors="replace")
            elif output_file and os.path.exists(output_file):
                with open(output_file, "rb") as fileobj:
                    detail = fileobj.read(200).decode(
                        "utf-8", errors="replace")
            suffix = ": %s" % detail if detail else ""
            raise RepoError("HTTP error %s%s" % (status, suffix))
        return body

    # -- Repo API ----------------------------------------------------------

    def itemlist(self, kind):
        """Returns a list of identifiers for each item of kind.

        Only pkgsinfo has a server-side list endpoint today; other kinds
        (catalogs, manifests, icons, pkgs) return an empty list so that
        makecatalogs and iconimporter degrade gracefully."""
        resource_type = kind.split("/", 1)[0]
        if resource_type != "pkgsinfo":
            return []
        body = self._request("pkgsinfo", accept="application/json")
        # Server response: {"items": [{"path": "Name/1.0.plist"}, ...]}
        decoded = json.loads(body)
        return [item["path"] for item in decoded.get("items", [])]

    def get(self, resource_identifier):
        """Returns the raw bytes of a repo item."""
        return self._request(resource_identifier)

    def get_to_local_file(self, resource_identifier, local_file_path):
        """Streams a repo item to local_file_path."""
        self._request(resource_identifier, output_file=local_file_path)

    def put(self, resource_identifier, content):
        """Uploads in-memory content to the repo. Use for pkginfo blobs."""
        with tempfile.NamedTemporaryFile(delete=False) as fileobj:
            fileobj.write(content)
            temp_path = fileobj.name
        try:
            self._request(
                resource_identifier, upload_file=temp_path,
                include_repo_id=True)
        finally:
            os.unlink(temp_path)

    def put_from_local_file(self, resource_identifier, local_file_path):
        """Streams local_file_path to the repo. Use for installer items."""
        self._request(
            resource_identifier, upload_file=local_file_path,
            include_repo_id=True)

    def delete(self, resource_identifier):
        """Deletes a repo item."""
        self._request(
            resource_identifier, method="DELETE", include_repo_id=True)
